src/ucp/dt/datatype_iter.inl

```diff
@@ -104,24 +104,78 @@ static UCS_F_ALWAYS_INLINE ucs_status_t ucp_datatype_iter_iov_allocate_memh(
     return UCS_OK;
 }
 
+static int
+ucp_datatype_iter_is_user_memh_valid(const ucp_datatype_iter_t *dt_iter,
+                                     const ucp_mem_h memh)
+{
+    ucs_string_buffer_t err_msg = UCS_STRING_BUFFER_INITIALIZER;
+    int is_valid                = 0;
+    int memh_contains_buffer;
+    size_t iov_count;
+
+    if (memh == NULL) {
+        ucs_error("got NULL memory handle");
+        goto out;
+    }
+
+    if (memh->mem_type != dt_iter->mem_info.type) {
+        ucs_error("unmatched memory type [%s]"
+                  " memh %p [%s]",
+                  ucs_memory_type_names[dt_iter->mem_info.type], memh,
+                  ucs_memory_type_names[memh->mem_type]);
+        goto out;
+    }
+
+    switch (dt_iter->dt_class) {
+    case UCP_DATATYPE_CONTIG:
+        memh_contains_buffer = ucp_memh_is_buffer_in_range(
```

Comment:
maybe use "goto err_memh_mismatch" instead of local variable memh_contains_buffer